{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barbara Karakyriakou\n",
    "## MCB112 - Fall 2018\n",
    "## homework 01:\n",
    "## the case of the dead sand mouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. check that the gene names match "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In principle, the gene names in the two data files ought to match, but (sigh) you know that all kinds of things can go wrong in practice: people use different names for the same gene, spell them with different capitalization or punctuation or added spaces, and so on. <br>Write a Python script to compare the gene names in the two data files. Output the names that appear in Moriarty_SuppTable1 but not Adler_SuppTable2, if any. <br>You're especially suspicious of the Moriarty et al. data table because he mentions in his paper that he exported the supplementary methods tables from Microsoft Excel, and you've run across people on Twitter discussing not just one but two terrifying papers about how Excel has systematically corrupted the supplementary data files in many published articles. <br> If there's a difference - why?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary from the Moriarty text file w/the gene names as keys and the measured data as values\n",
    "moriarty_data = {}                       \n",
    "for line in open('Moriarty_SuppTable1'):\n",
    "    if line[0] == '#': continue # skip header     \n",
    "    line   = line.rstrip('\\n') # remove the endings    \n",
    "    fields = line.split() # split each row into fields with space delimiter        \n",
    "\n",
    "    moriarty_data[fields[0]] = [float(s) for s in fields[1:6]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary from the Adler text file w/the gene names as keys and the measured data as values\n",
    "adler_data = {}                       \n",
    "for line in open('Adler_SuppTable2'):\n",
    "    if line[0] == '#': continue     \n",
    "    line   = line.rstrip('\\n')      \n",
    "    fields = line.split()           \n",
    "\n",
    "    adler_data[fields[0]] = [float(s) for s in fields[1:3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The unique genes for each table are: ['15-Sep', '2-Mar', '1-Mar', '10-Sep', '7-Mar', '4-Mar', '2-Sep', '11-Sep', '6-Mar', '11-Mar', '3-Mar', '8-Sep', '7-Sep', '14-Sep', '6-Sep', '1-Dec', '8-Mar', '5-Mar', '9-Mar', '12-Sep', '1-Sep', '4-Sep', '10-Mar', '9-Sep', '5-Sep', '3-Sep', 'MARCH4', 'MARCH1', 'SEPT10', 'MARCH8', 'SEP15', 'MARCH2', 'SEPT14', 'MARCH6', 'SEPT5', 'SEPT9', 'MARCH11', 'SEPT3', 'SEPT1', 'MARCH9', 'MARCH3', 'MARC2', 'SEPT8', 'MARCH7', 'SEPT7', 'MARCH5', 'MARC1', 'SEPT4', 'SEPT11', 'MARCH10', 'SEPT6', 'DEC1', 'SEPT12', 'SEPT2']\n"
     ]
    }
   ],
   "source": [
    "# list comprehension to find genes that are different in moriarty data\n",
    "diff1 = [gene for gene in moriarty_data if gene not in adler_data]\n",
    "# list comprehension to find genes that are different in adler data\n",
    "diff2 = [gene for gene in adler_data if gene not in moriarty_data]\n",
    "# combine the two lists\n",
    "diff = diff1+diff2\n",
    "print('The unique genes for each table are:', diff)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### There are differences in both files. It appears that some gene names erroneously converted to dates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. explore the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Write Python code to:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### output the five genes with the highest mRNA synthesis rate. (i.e. in Adler_SuppTable2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('LINC00176', [118.6, 13.8]),\n",
       " ('MAP3K9', [102.8, 14.2]),\n",
       " ('FTSJ1', [99.5, 15.6]),\n",
       " ('PIGP', [83.3, 17.7]),\n",
       " ('GDAP2', [73.9, 9.4])]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(adler_data.items(), key=lambda kv: kv[1], reverse=True)[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adler_data.items("
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### output the five genes with the longest mRNA halflife. (i.e. in Adler_SuppTable2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('BEX2', [3.6, 88.7]),\n",
       " ('LIPH', [7.9, 63.4]),\n",
       " ('NLK', [4.1, 62.9]),\n",
       " ('U2AF1', [9.6, 56.4]),\n",
       " ('TMEM186', [21.7, 53.9])]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(adler_data.items(), key=lambda kv: kv[1][1], reverse=True)[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### output the five genes that have the highest ratio of expression at t=96 hours post-mortem vs. t=0 (i.e. in Moriarty_SuppTable1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary with the ratio of expression t=96/t=0\n",
    "with open('Moriarty_SuppTable1') as f1:\n",
    "    ratio       = {}     \n",
    "    for line in f1:\n",
    "        if line[0] == '#': continue      \n",
    "        line   = line.rstrip('\\n')       \n",
    "        fields = line.split()            # field [1] = 0hr. [5] = 96hr.\n",
    "\n",
    "        ratio[fields[0]] = float(fields[5]) / float(fields[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('BEX2', 39.559064682660185),\n",
       " ('LIPH', 28.366606982990152),\n",
       " ('NLK', 27.083050207625522),\n",
       " ('U2AF1', 20.402556469970232),\n",
       " ('TMEM186', 19.621011673151752)]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# output the top five ratios\n",
    "sorted(ratio.items(), key=lambda kv: kv[1], reverse=True)[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. figure out what happened"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is partly an exercise in manipulating data files line-by-line in Python, but it's also designed to give you a Very Large Clue as to what happened in the dead sand mouse experiment.\n",
    "\n",
    "Write a Python script that merges the two data files, line by line, merging them on gene name. That is, for each line in file 1 for gene X, find the corresponding line for gene X in file 2; we're going to write a single output file with one line per gene. The genes are in different orders in the files, so this merge isn't entirely trivial. For any gene name X that isn't found in both files (cough cough Excel corruption cough cough) just skip it. For each gene name that is found in both files, output one whitespace-delimited, column-justified data line consisting of 7 fields per line:\n",
    "\n",
    "gene name\n",
    "Four expression ratios relative to t=0: i.e. tpm[12h]/tpm[0], tpm[24h]/tpm[0], tpm[48h]/tpm[0], tpm[96h]/tpm[0], by processing the TPM data in Moriarty_SuppTable1\n",
    "DNA synthesis rate (in mRNA/hr) and mRNA decay halflife (in hr) from Adler_SuppTable2\n",
    "Save your merged dataset to a file.\n",
    "\n",
    "Merging data sets, even crudely like this, is often useful as a step towards exploring data and looking for problems, outliers, and correlations.\n",
    "\n",
    "Explore the data, however you want, for example by looking at the genes with the highest expression ratio t=96/t=0. What do you think is the real explanation for what happened in the dead sand mouse experiment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Matplotlib is building the font cache using fc-list. This may take a moment.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>anise</td>\n",
       "      <td>71.1</td>\n",
       "      <td>36.5</td>\n",
       "      <td>15.6</td>\n",
       "      <td>3.1</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>apricot</td>\n",
       "      <td>28.2</td>\n",
       "      <td>31.0</td>\n",
       "      <td>36.1</td>\n",
       "      <td>27.6</td>\n",
       "      <td>11.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>artichoke</td>\n",
       "      <td>75.3</td>\n",
       "      <td>73.2</td>\n",
       "      <td>61.0</td>\n",
       "      <td>42.4</td>\n",
       "      <td>9.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>arugula</td>\n",
       "      <td>50.7</td>\n",
       "      <td>46.3</td>\n",
       "      <td>43.5</td>\n",
       "      <td>25.3</td>\n",
       "      <td>4.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>asparagus</td>\n",
       "      <td>17.3</td>\n",
       "      <td>4.5</td>\n",
       "      <td>1.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>avocado</td>\n",
       "      <td>26.2</td>\n",
       "      <td>20.8</td>\n",
       "      <td>17.2</td>\n",
       "      <td>7.5</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>banana</td>\n",
       "      <td>11.1</td>\n",
       "      <td>10.2</td>\n",
       "      <td>9.1</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>basil</td>\n",
       "      <td>46.3</td>\n",
       "      <td>48.0</td>\n",
       "      <td>44.2</td>\n",
       "      <td>28.0</td>\n",
       "      <td>8.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>beet</td>\n",
       "      <td>47.8</td>\n",
       "      <td>44.8</td>\n",
       "      <td>33.6</td>\n",
       "      <td>21.0</td>\n",
       "      <td>4.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>blackberry</td>\n",
       "      <td>16.5</td>\n",
       "      <td>1.3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name    0h   12h   24h   48h   96h\n",
       "0       anise  71.1  36.5  15.6   3.1   0.1\n",
       "1     apricot  28.2  31.0  36.1  27.6  11.9\n",
       "2   artichoke  75.3  73.2  61.0  42.4   9.7\n",
       "3     arugula  50.7  46.3  43.5  25.3   4.7\n",
       "4   asparagus  17.3   4.5   1.1   0.1   0.0\n",
       "5     avocado  26.2  20.8  17.2   7.5   0.8\n",
       "6      banana  11.1  10.2   9.1   4.5   0.7\n",
       "7       basil  46.3  48.0  44.2  28.0   8.9\n",
       "8        beet  47.8  44.8  33.6  21.0   4.4\n",
       "9  blackberry  16.5   1.3   0.1   0.0   0.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# convert the moriarty data to a dataframe and display the first 10 rows\n",
    "moriarty_df = pd.read_csv('Moriarty_SuppTable1', delim_whitespace=True, escapechar=\"#\")\n",
    "moriarty_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ZFY</td>\n",
       "      <td>7.2</td>\n",
       "      <td>24.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>FAM222B</td>\n",
       "      <td>0.5</td>\n",
       "      <td>15.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>MRPS33</td>\n",
       "      <td>0.9</td>\n",
       "      <td>25.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>LSM1</td>\n",
       "      <td>2.5</td>\n",
       "      <td>17.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>TIAL1</td>\n",
       "      <td>31.2</td>\n",
       "      <td>6.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>USP53</td>\n",
       "      <td>12.4</td>\n",
       "      <td>21.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>RP11-302M6.4</td>\n",
       "      <td>0.3</td>\n",
       "      <td>13.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>COLEC10</td>\n",
       "      <td>1.9</td>\n",
       "      <td>10.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>MTFR1L</td>\n",
       "      <td>1.5</td>\n",
       "      <td>9.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>SIPA1L1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>12.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name  synth_rate  halflife\n",
       "0           ZFY         7.2      24.3\n",
       "1       FAM222B         0.5      15.4\n",
       "2        MRPS33         0.9      25.4\n",
       "3          LSM1         2.5      17.4\n",
       "4         TIAL1        31.2       6.8\n",
       "5         USP53        12.4      21.8\n",
       "6  RP11-302M6.4         0.3      13.6\n",
       "7       COLEC10         1.9      10.1\n",
       "8        MTFR1L         1.5       9.9\n",
       "9       SIPA1L1         8.0      12.6"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert the adler data to a dataframe and display the first 10 rows\n",
    "adler_df = pd.read_csv('Adler_SuppTable2', delim_whitespace=True, escapechar=\"#\")\n",
    "adler_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>anise</td>\n",
       "      <td>71.1</td>\n",
       "      <td>36.5</td>\n",
       "      <td>15.6</td>\n",
       "      <td>3.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>8.2</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>apricot</td>\n",
       "      <td>28.2</td>\n",
       "      <td>31.0</td>\n",
       "      <td>36.1</td>\n",
       "      <td>27.6</td>\n",
       "      <td>11.9</td>\n",
       "      <td>1.7</td>\n",
       "      <td>17.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>artichoke</td>\n",
       "      <td>75.3</td>\n",
       "      <td>73.2</td>\n",
       "      <td>61.0</td>\n",
       "      <td>42.4</td>\n",
       "      <td>9.7</td>\n",
       "      <td>4.9</td>\n",
       "      <td>14.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>arugula</td>\n",
       "      <td>50.7</td>\n",
       "      <td>46.3</td>\n",
       "      <td>43.5</td>\n",
       "      <td>25.3</td>\n",
       "      <td>4.7</td>\n",
       "      <td>3.6</td>\n",
       "      <td>13.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>asparagus</td>\n",
       "      <td>17.3</td>\n",
       "      <td>4.5</td>\n",
       "      <td>1.1</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.1</td>\n",
       "      <td>5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>avocado</td>\n",
       "      <td>26.2</td>\n",
       "      <td>20.8</td>\n",
       "      <td>17.2</td>\n",
       "      <td>7.5</td>\n",
       "      <td>0.8</td>\n",
       "      <td>2.3</td>\n",
       "      <td>11.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>banana</td>\n",
       "      <td>11.1</td>\n",
       "      <td>10.2</td>\n",
       "      <td>9.1</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.7</td>\n",
       "      <td>0.9</td>\n",
       "      <td>12.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>basil</td>\n",
       "      <td>46.3</td>\n",
       "      <td>48.0</td>\n",
       "      <td>44.2</td>\n",
       "      <td>28.0</td>\n",
       "      <td>8.9</td>\n",
       "      <td>2.9</td>\n",
       "      <td>15.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>beet</td>\n",
       "      <td>47.8</td>\n",
       "      <td>44.8</td>\n",
       "      <td>33.6</td>\n",
       "      <td>21.0</td>\n",
       "      <td>4.4</td>\n",
       "      <td>3.4</td>\n",
       "      <td>13.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>blackberry</td>\n",
       "      <td>16.5</td>\n",
       "      <td>1.3</td>\n",
       "      <td>0.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.9</td>\n",
       "      <td>3.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name    0h   12h   24h   48h   96h  synth_rate  halflife\n",
       "0       anise  71.1  36.5  15.6   3.1   0.1         8.2       8.0\n",
       "1     apricot  28.2  31.0  36.1  27.6  11.9         1.7      17.1\n",
       "2   artichoke  75.3  73.2  61.0  42.4   9.7         4.9      14.3\n",
       "3     arugula  50.7  46.3  43.5  25.3   4.7         3.6      13.7\n",
       "4   asparagus  17.3   4.5   1.1   0.1   0.0         3.1       5.6\n",
       "5     avocado  26.2  20.8  17.2   7.5   0.8         2.3      11.6\n",
       "6      banana  11.1  10.2   9.1   4.5   0.7         0.9      12.7\n",
       "7       basil  46.3  48.0  44.2  28.0   8.9         2.9      15.1\n",
       "8        beet  47.8  44.8  33.6  21.0   4.4         3.4      13.3\n",
       "9  blackberry  16.5   1.3   0.1   0.0   0.0         3.9       3.7"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge the moriarty and adler dataframes ignoring the different gene names & display the first 10 rows\n",
    "merged_df = moriarty_df.merge(adler_df)\n",
    "merged_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# store data in a csv file at the current directory \n",
    "merged_df.to_csv('merged_data.csv', index = False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>15109</td>\n",
       "      <td>TMEM186</td>\n",
       "      <td>1285.0</td>\n",
       "      <td>2154.7</td>\n",
       "      <td>4079.0</td>\n",
       "      <td>8581.4</td>\n",
       "      <td>25213.0</td>\n",
       "      <td>21.7</td>\n",
       "      <td>53.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12958</td>\n",
       "      <td>CAND1</td>\n",
       "      <td>1039.7</td>\n",
       "      <td>1815.3</td>\n",
       "      <td>2615.8</td>\n",
       "      <td>5888.8</td>\n",
       "      <td>13543.7</td>\n",
       "      <td>21.4</td>\n",
       "      <td>45.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8527</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>329.3</td>\n",
       "      <td>600.2</td>\n",
       "      <td>1292.2</td>\n",
       "      <td>3126.9</td>\n",
       "      <td>13026.8</td>\n",
       "      <td>3.6</td>\n",
       "      <td>88.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4380</td>\n",
       "      <td>LIPH</td>\n",
       "      <td>446.8</td>\n",
       "      <td>1071.2</td>\n",
       "      <td>1748.4</td>\n",
       "      <td>3692.4</td>\n",
       "      <td>12674.2</td>\n",
       "      <td>7.9</td>\n",
       "      <td>63.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2867</td>\n",
       "      <td>GTDC1</td>\n",
       "      <td>960.3</td>\n",
       "      <td>1724.1</td>\n",
       "      <td>2769.7</td>\n",
       "      <td>5484.0</td>\n",
       "      <td>12031.2</td>\n",
       "      <td>21.1</td>\n",
       "      <td>44.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19903</td>\n",
       "      <td>U2AF1</td>\n",
       "      <td>571.1</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>1810.9</td>\n",
       "      <td>3913.0</td>\n",
       "      <td>11651.9</td>\n",
       "      <td>9.6</td>\n",
       "      <td>56.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5267</td>\n",
       "      <td>TRIO</td>\n",
       "      <td>1079.2</td>\n",
       "      <td>1753.1</td>\n",
       "      <td>2384.1</td>\n",
       "      <td>4889.0</td>\n",
       "      <td>8731.7</td>\n",
       "      <td>28.1</td>\n",
       "      <td>36.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4346</td>\n",
       "      <td>LAMP3</td>\n",
       "      <td>1564.1</td>\n",
       "      <td>2399.7</td>\n",
       "      <td>3202.6</td>\n",
       "      <td>5083.1</td>\n",
       "      <td>8526.4</td>\n",
       "      <td>48.8</td>\n",
       "      <td>30.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>218</td>\n",
       "      <td>SLC25A33</td>\n",
       "      <td>735.6</td>\n",
       "      <td>1167.7</td>\n",
       "      <td>1746.3</td>\n",
       "      <td>3131.9</td>\n",
       "      <td>8204.5</td>\n",
       "      <td>16.7</td>\n",
       "      <td>40.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16124</td>\n",
       "      <td>NLK</td>\n",
       "      <td>264.9</td>\n",
       "      <td>492.4</td>\n",
       "      <td>945.4</td>\n",
       "      <td>2000.6</td>\n",
       "      <td>7174.3</td>\n",
       "      <td>4.1</td>\n",
       "      <td>62.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name      0h     12h     24h     48h      96h  synth_rate  \\\n",
       "15109    TMEM186  1285.0  2154.7  4079.0  8581.4  25213.0        21.7   \n",
       "12958      CAND1  1039.7  1815.3  2615.8  5888.8  13543.7        21.4   \n",
       "8527        BEX2   329.3   600.2  1292.2  3126.9  13026.8         3.6   \n",
       "4380        LIPH   446.8  1071.2  1748.4  3692.4  12674.2         7.9   \n",
       "2867       GTDC1   960.3  1724.1  2769.7  5484.0  12031.2        21.1   \n",
       "19903      U2AF1   571.1  1015.0  1810.9  3913.0  11651.9         9.6   \n",
       "5267        TRIO  1079.2  1753.1  2384.1  4889.0   8731.7        28.1   \n",
       "4346       LAMP3  1564.1  2399.7  3202.6  5083.1   8526.4        48.8   \n",
       "218     SLC25A33   735.6  1167.7  1746.3  3131.9   8204.5        16.7   \n",
       "16124        NLK   264.9   492.4   945.4  2000.6   7174.3         4.1   \n",
       "\n",
       "       halflife  \n",
       "15109      53.9  \n",
       "12958      45.3  \n",
       "8527       88.7  \n",
       "4380       63.4  \n",
       "2867       44.2  \n",
       "19903      56.4  \n",
       "5267       36.5  \n",
       "4346       30.9  \n",
       "218        40.2  \n",
       "16124      62.9  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore data sorting by the genes with the highest t=96/t=0 expression ratio\n",
    "(merged_df.sort_values(by=['96h'], ascending=False)).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>8527</td>\n",
       "      <td>BEX2</td>\n",
       "      <td>329.3</td>\n",
       "      <td>600.2</td>\n",
       "      <td>1292.2</td>\n",
       "      <td>3126.9</td>\n",
       "      <td>13026.8</td>\n",
       "      <td>3.6</td>\n",
       "      <td>88.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4380</td>\n",
       "      <td>LIPH</td>\n",
       "      <td>446.8</td>\n",
       "      <td>1071.2</td>\n",
       "      <td>1748.4</td>\n",
       "      <td>3692.4</td>\n",
       "      <td>12674.2</td>\n",
       "      <td>7.9</td>\n",
       "      <td>63.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16124</td>\n",
       "      <td>NLK</td>\n",
       "      <td>264.9</td>\n",
       "      <td>492.4</td>\n",
       "      <td>945.4</td>\n",
       "      <td>2000.6</td>\n",
       "      <td>7174.3</td>\n",
       "      <td>4.1</td>\n",
       "      <td>62.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19903</td>\n",
       "      <td>U2AF1</td>\n",
       "      <td>571.1</td>\n",
       "      <td>1015.0</td>\n",
       "      <td>1810.9</td>\n",
       "      <td>3913.0</td>\n",
       "      <td>11651.9</td>\n",
       "      <td>9.6</td>\n",
       "      <td>56.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15109</td>\n",
       "      <td>TMEM186</td>\n",
       "      <td>1285.0</td>\n",
       "      <td>2154.7</td>\n",
       "      <td>4079.0</td>\n",
       "      <td>8581.4</td>\n",
       "      <td>25213.0</td>\n",
       "      <td>21.7</td>\n",
       "      <td>53.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16598</td>\n",
       "      <td>CHAD</td>\n",
       "      <td>105.0</td>\n",
       "      <td>202.1</td>\n",
       "      <td>339.5</td>\n",
       "      <td>695.6</td>\n",
       "      <td>2037.5</td>\n",
       "      <td>1.9</td>\n",
       "      <td>53.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17923</td>\n",
       "      <td>LRG1</td>\n",
       "      <td>41.6</td>\n",
       "      <td>83.4</td>\n",
       "      <td>132.5</td>\n",
       "      <td>303.9</td>\n",
       "      <td>752.8</td>\n",
       "      <td>0.8</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6302</td>\n",
       "      <td>OR14J1</td>\n",
       "      <td>141.9</td>\n",
       "      <td>267.5</td>\n",
       "      <td>473.9</td>\n",
       "      <td>984.4</td>\n",
       "      <td>2521.7</td>\n",
       "      <td>2.9</td>\n",
       "      <td>50.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1273</td>\n",
       "      <td>CDC42SE1</td>\n",
       "      <td>33.6</td>\n",
       "      <td>65.6</td>\n",
       "      <td>103.4</td>\n",
       "      <td>230.5</td>\n",
       "      <td>593.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>49.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13269</td>\n",
       "      <td>TMEM120B</td>\n",
       "      <td>82.1</td>\n",
       "      <td>166.3</td>\n",
       "      <td>253.7</td>\n",
       "      <td>524.2</td>\n",
       "      <td>1440.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>48.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name      0h     12h     24h     48h      96h  synth_rate  \\\n",
       "8527        BEX2   329.3   600.2  1292.2  3126.9  13026.8         3.6   \n",
       "4380        LIPH   446.8  1071.2  1748.4  3692.4  12674.2         7.9   \n",
       "16124        NLK   264.9   492.4   945.4  2000.6   7174.3         4.1   \n",
       "19903      U2AF1   571.1  1015.0  1810.9  3913.0  11651.9         9.6   \n",
       "15109    TMEM186  1285.0  2154.7  4079.0  8581.4  25213.0        21.7   \n",
       "16598       CHAD   105.0   202.1   339.5   695.6   2037.5         1.9   \n",
       "17923       LRG1    41.6    83.4   132.5   303.9    752.8         0.8   \n",
       "6302      OR14J1   141.9   267.5   473.9   984.4   2521.7         2.9   \n",
       "1273    CDC42SE1    33.6    65.6   103.4   230.5    593.4         0.6   \n",
       "13269   TMEM120B    82.1   166.3   253.7   524.2   1440.8         1.7   \n",
       "\n",
       "       halflife  \n",
       "8527       88.7  \n",
       "4380       63.4  \n",
       "16124      62.9  \n",
       "19903      56.4  \n",
       "15109      53.9  \n",
       "16598      53.5  \n",
       "17923      52.0  \n",
       "6302       50.4  \n",
       "1273       49.6  \n",
       "13269      48.7  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore data sorting by the genes with the highest halflife\n",
    "(merged_df.sort_values(by=['halflife'], ascending=False)).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>17764</td>\n",
       "      <td>LINC00176</td>\n",
       "      <td>1615.0</td>\n",
       "      <td>1558.6</td>\n",
       "      <td>1488.1</td>\n",
       "      <td>861.4</td>\n",
       "      <td>185.1</td>\n",
       "      <td>118.6</td>\n",
       "      <td>13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14067</td>\n",
       "      <td>MAP3K9</td>\n",
       "      <td>1489.0</td>\n",
       "      <td>1509.3</td>\n",
       "      <td>1412.0</td>\n",
       "      <td>963.7</td>\n",
       "      <td>217.7</td>\n",
       "      <td>102.8</td>\n",
       "      <td>14.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8247</td>\n",
       "      <td>FTSJ1</td>\n",
       "      <td>1582.1</td>\n",
       "      <td>1794.5</td>\n",
       "      <td>1553.1</td>\n",
       "      <td>1141.6</td>\n",
       "      <td>417.8</td>\n",
       "      <td>99.5</td>\n",
       "      <td>15.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19860</td>\n",
       "      <td>PIGP</td>\n",
       "      <td>1501.1</td>\n",
       "      <td>1865.1</td>\n",
       "      <td>1933.2</td>\n",
       "      <td>1678.4</td>\n",
       "      <td>801.9</td>\n",
       "      <td>83.3</td>\n",
       "      <td>17.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1157</td>\n",
       "      <td>GDAP2</td>\n",
       "      <td>685.8</td>\n",
       "      <td>477.2</td>\n",
       "      <td>280.2</td>\n",
       "      <td>70.4</td>\n",
       "      <td>3.1</td>\n",
       "      <td>73.9</td>\n",
       "      <td>9.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14033</td>\n",
       "      <td>ATP6V1D</td>\n",
       "      <td>1655.8</td>\n",
       "      <td>2452.9</td>\n",
       "      <td>3396.0</td>\n",
       "      <td>4211.4</td>\n",
       "      <td>5210.0</td>\n",
       "      <td>64.4</td>\n",
       "      <td>26.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9715</td>\n",
       "      <td>SIT1</td>\n",
       "      <td>439.7</td>\n",
       "      <td>174.6</td>\n",
       "      <td>67.4</td>\n",
       "      <td>6.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>64.1</td>\n",
       "      <td>6.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1006</td>\n",
       "      <td>PTBP2</td>\n",
       "      <td>802.3</td>\n",
       "      <td>845.4</td>\n",
       "      <td>740.3</td>\n",
       "      <td>450.6</td>\n",
       "      <td>100.2</td>\n",
       "      <td>60.2</td>\n",
       "      <td>13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12209</td>\n",
       "      <td>ADRB1</td>\n",
       "      <td>575.7</td>\n",
       "      <td>381.3</td>\n",
       "      <td>242.8</td>\n",
       "      <td>60.1</td>\n",
       "      <td>3.1</td>\n",
       "      <td>59.5</td>\n",
       "      <td>9.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17397</td>\n",
       "      <td>CST1</td>\n",
       "      <td>1347.8</td>\n",
       "      <td>1604.5</td>\n",
       "      <td>2297.3</td>\n",
       "      <td>2323.6</td>\n",
       "      <td>1991.0</td>\n",
       "      <td>58.5</td>\n",
       "      <td>22.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name      0h     12h     24h     48h     96h  synth_rate  halflife\n",
       "17764  LINC00176  1615.0  1558.6  1488.1   861.4   185.1       118.6      13.8\n",
       "14067     MAP3K9  1489.0  1509.3  1412.0   963.7   217.7       102.8      14.2\n",
       "8247       FTSJ1  1582.1  1794.5  1553.1  1141.6   417.8        99.5      15.6\n",
       "19860       PIGP  1501.1  1865.1  1933.2  1678.4   801.9        83.3      17.7\n",
       "1157       GDAP2   685.8   477.2   280.2    70.4     3.1        73.9       9.4\n",
       "14033    ATP6V1D  1655.8  2452.9  3396.0  4211.4  5210.0        64.4      26.1\n",
       "9715        SIT1   439.7   174.6    67.4     6.6     0.0        64.1       6.9\n",
       "1006       PTBP2   802.3   845.4   740.3   450.6   100.2        60.2      13.8\n",
       "12209      ADRB1   575.7   381.3   242.8    60.1     3.1        59.5       9.5\n",
       "17397       CST1  1347.8  1604.5  2297.3  2323.6  1991.0        58.5      22.1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# explore data sorting by the genes with the highest synthesis rate\n",
    "(merged_df.sort_values(by=['synth_rate'], ascending=False)).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *I will filter the data to check each gene of the experiment diagrams.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>19586</td>\n",
       "      <td>APOBEC3G</td>\n",
       "      <td>142.8</td>\n",
       "      <td>198.9</td>\n",
       "      <td>230.9</td>\n",
       "      <td>204.5</td>\n",
       "      <td>112.2</td>\n",
       "      <td>8.4</td>\n",
       "      <td>18.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name     0h    12h    24h    48h    96h  synth_rate  halflife\n",
       "19586   APOBEC3G  142.8  198.9  230.9  204.5  112.2         8.4      18.8"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filter data by gene name\n",
    "merged_df[merged_df.values  == ('APOBEC3G')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* 'APOBEC3G' appears like it continues being expressed after the mouse is dead, but in reality this is due to slow decay of the mRNA. RNA-seq measures the relative abundance of each transcript, not the absolute abundance, so it looks like the \"expression\" of the more stable mRNAs is going up, just because they're going away slower. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>24</td>\n",
       "      <td>cilantro</td>\n",
       "      <td>394.3</td>\n",
       "      <td>294.7</td>\n",
       "      <td>184.3</td>\n",
       "      <td>62.2</td>\n",
       "      <td>3.7</td>\n",
       "      <td>37.7</td>\n",
       "      <td>10.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name     0h    12h    24h   48h  96h  synth_rate  halflife\n",
       "24   cilantro  394.3  294.7  184.3  62.2  3.7        37.7      10.3"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df.values  == ('cilantro')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* 'cilantro's' expression drops immediately after the mouse dies, which means that this mRNA is not stable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>16781</td>\n",
       "      <td>GRIN2C</td>\n",
       "      <td>210.6</td>\n",
       "      <td>260.8</td>\n",
       "      <td>306.0</td>\n",
       "      <td>401.3</td>\n",
       "      <td>233.7</td>\n",
       "      <td>9.8</td>\n",
       "      <td>20.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name     0h    12h    24h    48h    96h  synth_rate  halflife\n",
       "16781     GRIN2C  210.6  260.8  306.0  401.3  233.7         9.8      20.8"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df.values  == ('GRIN2C')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* 'GRIN2C' similar to 'APOBEC3G' appears as being expressed due to existing mRNA that decays slowly after the mouse is dead. An mRNA with a moderate halflife will go up (as the least stable mRNAs disappear quickly from the population), then go down (as the moderate-halflife mRNA decays faster than more stable mRNAs)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>0h</th>\n",
       "      <th>12h</th>\n",
       "      <th>24h</th>\n",
       "      <th>48h</th>\n",
       "      <th>96h</th>\n",
       "      <th>synth_rate</th>\n",
       "      <th>halflife</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>13269</td>\n",
       "      <td>TMEM120B</td>\n",
       "      <td>82.1</td>\n",
       "      <td>166.3</td>\n",
       "      <td>253.7</td>\n",
       "      <td>524.2</td>\n",
       "      <td>1440.8</td>\n",
       "      <td>1.7</td>\n",
       "      <td>48.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name    0h    12h    24h    48h     96h  synth_rate  halflife\n",
       "13269   TMEM120B  82.1  166.3  253.7  524.2  1440.8         1.7      48.7"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[merged_df.values  == ('TMEM120B')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* 'TMEM120B' is also similar to 'GRIN2C' similar to 'APOBEC3G. Overall, we see that the least stable mRNAs go away fast, while the more stable mRNAs decay in a slower rate. mRNAs with higher halflife appear as being expressed after death, but this is not true. As noted above, RNA-seq measures the relative abundance of each transcript, not the absolute abundance, so it looks like the \"expression\" of the more stable mRNAs is going up,"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
